### PR TITLE
Use Cardano Haskell package repository ("CHaP")

### PR DESCRIPTION
**Pull request by andreabedini**
This PR replaces most `source-repository-package` stanzas in cabal.project with packages fetched from our [Cardano Haskell package repository](https://input-output-hk.github.io/cardano-haskell-packages/).

Unfortunately some packages in CHaP still have broken version bounds so we need to manually specify a bunch of constraints to make sure we have a plan we can build.

I still need to revisit the documentation in `CONTRIBUTING`.

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -24,6 +41,86 @@ packages: cardano-streaming
...
+  , cardano-prelude                  == 0.1.0.0
+
+  -- needed by cardano-ledger-byron-0.1.0.0
+  , cardano-binary                   == 1.5.0
+
+  -- needed by cardano-ledger-core-0.1.0.0
+  , cardano-crypto-class             == 2.0.0
+
+  --
+  -- the following constraints are speculative and reflect the current build plan
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply cabal.project:62]: some of these constraints should be in the cabal files in the repo, some should be in other cabal files. Unfortunately at this point I cannot tell which ones are which. :shrug:

**Comment by andreabedini** [comment]
the failure doesn't seem to have anything to do with CHaP

```
unpacking sources
unpacking source archive /nix/store/0fhy2yp7avaaxcvvk7iw9y4zz9zzxbsp-plutus-apps-root-playground-common-test-playground-common-test-root
source root is plutus-apps-root-playground-common-test-playground-common-test-root
patching sources
building
all tests
  Auth.TypesSpec
    OAuthToken JSON handling
      should be able to parse a github response: FAIL
        Exception: /nix/store/swckk3qwxhhii52wird3gb61698gwc4l-playground-common-test-playground-common-test-1.0.0.0-data/share/ghc-8.10.7/x86_64-linux-ghc-8.10.7/playground-common-1.0.0.0/test/oAuthToken1.json: openBinaryFile: does not exist (No such file or directory)
        Use -p '/should be able to parse a github response/' to rerun this test only.
  Schema
    toSchema
      Encoding of various types:                 OK
    toArgument
      Encoding of various types:                 OK
  TH
    Schemas compile as expected:                 OK
  mkKnownCurrencies
    Serialisation:                               OK

1 out of 5 tests failed (0.04s)```

File: nix/pkgs/haskell/haskell.nix [hunk 2]
```diff
@@ -26,6 +27,9 @@ let
         name = "plutus-apps";
       };
     sha256map = import ./sha256map.nix;
```
Comment (zeme-iohk) [inline nix/pkgs/haskell/haskell.nix:29]: Do we need this or can we do `sha256Map = {}`? (<-- we do this in `plutus`)

Comment (andreabedini) [reply nix/pkgs/haskell/haskell.nix:29]: `plutus` managed to get rid of all `source-repository-package`s, so they don't need `sha256Map` anymore. That said I should remove from `nix/pkgs/haskell/sha256map.nix` the hashes that are not needed anymore. Good catch.

**Comment by koslambrou** [comment]
> the failure doesn't seem to have anything to do with CHaP

That's the same error as https://github.com/input-output-hk/plutus-apps/pull/757 which tries to bump haskell.nix and hackage.nix. Is using CHaP making us use a more recent version of those packages?

**Comment by michaelpj** [comment]
Andrea just bumped `haskell.nix` in this PR also, so it probably has the same bug as that one. Needs a newer haskell.nix. Ideally we just fix that other PR soon!

**Comment by andreabedini** [comment]
WUT? I force pushed and GitHub closed the PR. :shrug: I'll reopen a new one if I can't reopen it.

File: cabal.project [hunk 3]
```diff
@@ -156,58 +268,14 @@ source-repository-package
...
-
 -- Should follow cardano-node
 source-repository-package
   type: git
   location: https://github.com/input-output-hk/cardano-config
   tag: 1646e9167fab36c0bff82317743b96efa2d3adaa
 
--- Should follow cardano-wallet.
+-- This is needed because we rely on an unreleased feature
+-- https://github.com/input-output-hk/cardano-ledger/pull/3111
```
Comment (michaelpj) [inline cabal.project:278]: :see_no_evil:

File: cabal.project [hunk 4]
```diff
@@ -156,58 +268,14 @@ source-repository-package
...
-    stubs/plutus-ghc-stub
-    word-array
-
--- Should follow cardano-node
-source-repository-package
-  type: git
-  location: https://github.com/input-output-hk/ekg-forward
-  tag: 297cd9db5074339a2fb2e5ae7d0780debb670c63
-
 -- Should follow cardano-node
```
Comment (koslambrou) [inline cabal.project:271]: If this one wasn't removed, that's because it was not released on CHaP?

File: cabal.project [hunk 5]
```diff
@@ -48,6 +65,113 @@ constraints:
   , protolude < 0.3.1
   , algebraic-graphs < 0.7
 
+constraints:
```
Comment (koslambrou) [inline cabal.project:68]: I would add a comment here explaining why we need those, and what needs to be done to be able to remove them in the future.

File: cabal.project [hunk 6]
```diff
@@ -1,6 +1,23 @@
--- Bump this if you need newer packages.
+-- Custom repository for cardano haskell packages
```
Comment (koslambrou) [inline cabal.project:1]: Can we add the following?

See https://github.com/input-output-hk/cardano-haskell-packages on how to use CHaP in a Haskell project.

File: cabal.project (hunk 4 above)
Comment (andreabedini) [reply cabal.project:271]: It was removed. The exact same version is present in CHaP https://input-output-hk.github.io/cardano-haskell-packages/package/ekg-forward-0.1.0/

File: cabal.project (hunk 5 above)
Comment (andreabedini) [reply cabal.project:68]: Yes, I will do it. We did something similar for the other repositories.

File: cabal.project (hunk 6 above)
Comment (andreabedini) [reply cabal.project:1]: :+1:

File: cabal.project (hunk 4 above)
Comment (koslambrou) [reply cabal.project:271]: I'm actually referring to the line below: `cardano-config`

Comment (andreabedini) [reply cabal.project:271]: Ah, yes, cardano-config is not in CHaP. I notice it's marked as deprecated and we don't actually seem to use it anywhere. I'll remove it.

Comment (andreabedini) [reply cabal.project:271]: (it was once a dependecy of cardano-node but it has not been since January)

File: cabal.project (hunk 3 above)
Comment (andreabedini) [reply cabal.project:278]: Adding insult to the injury, cardano-ledger just got a new release out and that feature is not included :sob:

File: CONTRIBUTING.adoc [hunk 7]
```diff
@@ -108,55 +108,78 @@ NOTE: you must run your editor from the nix shell, and not from, say, an
...
-actually use flakes for normal Nix builds. To manage the dependencies,
-see https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-flake.html#flake-inputs[the Nix flake inputs documentation]
-and https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-flake-lock.html[the Nix flake lock command].
+You can use a ``source-repository-package`` stanza to pull in the
+unreleased version.
 
-Specifically, you will probabl want to say `nix flake lock --update-input <input-name>`.
-Do *not* use `nix flake update`, as that will update all the inputs, which we typically don't want to do.
+Try only to do this for a short time, as it does not play very well with
+tooling, and will interfere with the ability to release the node itself.
```
Comment (koslambrou) [inline CONTRIBUTING.adoc:155]: I don't think that's relevant for plutus-apps

File: CONTRIBUTING.adoc [hunk 8]
```diff
@@ -108,55 +108,78 @@ NOTE: you must run your editor from the nix shell, and not from, say, an
...
-Note that `cabal` itself keeps track of what index states it knows about, so you may need to update this with `cabal update` in order for `cabal` to be happy.
+See the README for instructions, including a script to automate most of the
+process.
 
-The Nix code which builds our packages also cares about the index state.
-The set of index states which it knows about is controlled by `hackage.nix`, which is a Nix representation of Hackage.
-This therefore needs to be newer than the index state.
-You can update it xref:update-nix-pins[with the Nix flake commands].
+Please note that libraries need bounds on the version of their dependencies
+to avoid bitrot and be effectively reusable.
```
Comment (koslambrou) [inline CONTRIBUTING.adoc:182]: We have a doc about release process: `doc/BRANCHING-AND-RELEASE.adoc`.

It should be updated with this.

File: CONTRIBUTING.adoc (hunk 7 above)
Comment (andreabedini) [reply CONTRIBUTING.adoc:155]: Sorry bad copy & paste. Fixed

File: CONTRIBUTING.adoc (hunk 8 above)
Comment (andreabedini) [reply CONTRIBUTING.adoc]: Thanks, done. I removed that part from CONTRIBUTING and added a sentence to doc/BRANCHING-AND-RELEASE.adoc

**Comment by andreabedini** [comment]
I had to reintroduce a srp for cardano-node because `cardano-git-rev cardano-submit-api trace-dispatcher trace-resources trace-forward` were not released to chap along with cardano-node 1.35.4. Hopefully this is only temporary.

**Comment by michaelpj** [comment]
> I had to reintroduce a srp for cardano-node because cardano-git-rev cardano-submit-api trace-dispatcher trace-resources trace-forward were not released to chap along with cardano-node 1.35.4.

Are those packages that _we_ depend on, or ones that the released packages depend on? We could probably release those now if they should have been released.

**Comment by michaelpj** [comment]
Can we move to the Hackage `quickcheck-dynamic` now, perhaps?

**Comment by andreabedini** [comment]
@michaelpj 
> Are those packages that _we_ depend on, or ones that the released packages depend on? We could probably release those now if they should have been released.

we use cardano-submit-api in plutus-example, the other ones are transitive. We [just released](https://github.com/input-output-hk/cardano-haskell-packages/pull/100) trace-dispatcher and trace-forward which are the critical ones.

**Comment by andreabedini** [comment]
> Can we move to the Hackage `quickcheck-dynamic` now, perhaps?

yes but it's not entirely trivial, we have ticket for that PLT-1239

**Comment by andreabedini** [comment]
I think this is the best we can do for the moment.

**Review by michaelpj — APPROVED** [review]
We'll need to resolve the ledger issue before we can publish plutus-apps libs to CHaP, and also there's going to be a lot of local recompilation for folks still until we do, but I agree this is the best we can do for now :+1:

File: CONTRIBUTING.adoc [hunk 9]
```diff
@@ -108,55 +108,61 @@ NOTE: you must run your editor from the nix shell, and not from, say, an
...
+Updating package dependencies from Hackage should work like normal in a
+Haskell project. The most important thing to note is that we pin the
+``index-state`` of the Hackage package index in ``cabal.project``. This
+means that cabal will always see Hackage “as if” it was that time, ensuring
+reproducibility. But it also means that if you need a package version that
+was released *after* that time, you need to bump the ``index-state`` (and
+to run ``cabal update`` locally). Please also note that index-state is a
+property of your working environment (the cabal project) not of the
+packages. This means that downstream consumers have no idea of your
+`index-state` setting and you need to make sure your packages work
```
Comment (zeme-iohk) [inline CONTRIBUTING.adoc:125]: nit: single ` as opposed to double `` here and on line 126

**Review by zeme-iohk — APPROVED** [review]

File: cabal.project (hunk 3 above)
Comment (koslambrou) [reply cabal.project:188]: Well, their new release is based on functionality that was done on August. Not really up to date. Don't know why though.

**Review by koslambrou — APPROVED** [review]

File: CONTRIBUTING.adoc (hunk 9 above)
Comment (andreabedini) [reply CONTRIBUTING.adoc]: :pray:

File: cabal.project [hunk 10]
```diff
@@ -35,18 +53,37 @@ benchmarks: true
...
-  , size-based:template-haskell
+  -- this is material for the hackage trustees
+  , ekg:aeson
+  , ekg-json:aeson
+  , openapi3:aeson
+  , servant:aeson
+  , servant-client-core:aeson
+  , servant-server:aeson
 
 constraints:
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply cabal.project:68]: Done.

File: cabal.project [hunk 11]
```diff
@@ -35,18 +53,37 @@ benchmarks: true
 -- 'tasty' output.
 test-show-details: direct
 
--- cardano-addresses unit tests bring in some version constraint conflicts:
---
--- * it has strict aeson < 1.5 dep - this will be fixed in the next release.
 allow-newer:
-    *:aeson
-  , size-based:template-haskell
+  -- this is material for the hackage trustees
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply cabal.project]: Added a comment.

**Pull request merged**